compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java

```diff
@@ -164,62 +125,100 @@ private STToken readTokenInReTerm() {
         int nextChar = peek();
         switch (nextChar) {
             case LexerTerminals.BITWISE_XOR:
+                this.reader.advance();
+                return getRegExpText(SyntaxKind.RE_ASSERTION_VALUE);
             case LexerTerminals.DOLLAR:
                 this.reader.advance();
+                if (peek() == LexerTerminals.OPEN_BRACE) {
+                    // Start interpolation mode.
+                    startMode(ParserMode.INTERPOLATION);
+                    this.reader.advance();
+                    return getRegExpSyntaxToken(SyntaxKind.INTERPOLATION_START_TOKEN);
+                }
                 return getRegExpText(SyntaxKind.RE_ASSERTION_VALUE);
             case LexerTerminals.DOT:
                 this.reader.advance();
                 startMode(ParserMode.RE_QUANTIFIER);
```

Comment:
Since `.` is just a syntax char and we don't need to change the way we tokenize, we don't need to start a mode here. Basically, every syntax char should have a `case` in this `switch` ( I c a few missing eg `+, *, }, ]`). We need to only start the modes when we need to change the way we tokenize. eg: by seeing `{` we have to start a mode since we need to identify `sc=`, `gc=`,  `AsciiLetter`,  `Digit `, `_`, `,`